sklearn/ensemble/tests/test_gradient_boosting_loss_functions.py

```diff
@@ -184,3 +185,49 @@ def test_sample_weight_deviance():
         deviance_w_w = loss(y, p, sample_weight)
         deviance_wo_w = loss(y, p)
         assert deviance_wo_w == deviance_w_w
+
+
+def test_multinomial_deviance():
+    # Check multinomial deviance with and without sample weights.
+    rng = check_random_state(13)
+    sample_weight = np.ones(100)
+    y = rng.randint(0, 3, size=sample_weight.shape[0])
+    p = np.zeros((y.shape[0], 3), dtype=np.float64)
+    for i in range(p.shape[1]):
+        p[:, i] = y == i
+
+    loss = MultinomialDeviance(3)
+    loss_wo_sw = loss(y, p)
+    assert loss_wo_sw > 0
+    loss_w_sw = loss(y, p, np.ones(p.shape[0], dtype=np.float32))
+    assert_almost_equal(loss_wo_sw, loss_w_sw)
+    loss_w_sw = loss(y, p, 0.5 * np.ones(p.shape[0], dtype=np.float32))
+    assert_almost_equal(loss_wo_sw, loss_w_sw)
+
+
+def test_mdl_computation_unweighted(pred=np.array([[1.0, 0, 0],
+                                                   [0, 0.5, 0.5]]),
+                                    y=np.array([0, 1]),
+                                    expected_loss=0.75473):
+    # MultinomialDeviance loss computation with uniform/without weights.
+    weights = np.array([1, 1])
+    loss = MultinomialDeviance(3)
+    assert_almost_equal(loss(y, pred, weights), expected_loss, decimal=4)
+    assert_almost_equal(loss(y, pred, None), expected_loss, decimal=4)
+    assert_almost_equal(loss(y, pred), expected_loss, decimal=4)
+
+
+def test_mdl_computation_weighted(pred=np.array([[1.0, 0, 0],
+                                                 [0, 0.5, 0.5]]),
+                                  y=np.array([0, 1]),
+                                  weights=np.array([1, 3]),
+                                  expected_loss=0.85637):
+    # MultinomialDeviance loss computation with weights.
+    loss = MultinomialDeviance(3)
+    assert_almost_equal(loss(y, pred, weights), expected_loss, decimal=4)
+
+
+def test_mdl_exception():
+    # Check that MultinomialDeviance throws when n_classes <= 2
+    assert_raises(ValueError, MultinomialDeviance, 2)
```

Comment:
Use assert_raises_regex and check which message is raise. We tend to check the error message rather than only checking that something is wrong ;)

Reply:
Shall we use the context manager here?

Reply:
I would not since sklearn already have helper function. If we want to use the context manager, we need to change all other asser_raises_*